### ISSUE TITLE
NotFoundHttpException when Route URLs are the same.

### ISSUE THREAD

**Issue by greentaxi-com** [skipperbent/simple-php-router#492]
Hi,

I have four routes and when I call any route I am getting a Pecee\SimpleRouter\Exceptions\NotFoundHttpException - Route "/supply/a49503dd-be09-4641-92e5-a079b11ff367/" or method "post" not allowed.

    Router::post('/supply/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->CreateSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);
    
    Router::get('/supply/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->GetSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);
    
    Router::put('/supply/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->UpdateSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);
    
    Router::delete('/supply/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->DeleteSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);

If I change the Urls to the following, it works with no exception. Why? 

    Router::post('/supply_post/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->CreateSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);
    
    Router::get('/supply_get/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->GetSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);
    
    Router::put('/supply_put/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->UpdateSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);
    
    Router::delete('/supply_delete/{operatorUUID?}/{supplyUUID?}/', function ($operatorUUID = Null, $supplyUUID = Null) { /* Authentication::Authorise($operatorUUID); */
        $supply = new Supply($operatorUUID, $supplyUUID);
        $supply->DeleteSupply();
    }, ['defaultParameterRegex' => '[\w\-]+']);

**Comment by skipperbent** [comment]
I can see how those routes would cause an error. The routes will match, even when the request-method is not allowed depending on what order they're added - causing an error. The router will not check if other routes exists with the same exact url, it will just throw an error right then and there.

You might want to look into the `resource` option instead and create a `Controller` that implements the `IResourceController`.

**Route example:**

`Router::resource('/demo', 'ApiController');`

**Controller example:**

https://github.com/skipperbent/simple-router-demo/blob/master/app/Controllers/ApiController.php

**Comment by skipperbent** [comment]
Hi.

The next version (version 4.3) will allow you to add multiple routes with different request-types.
It's not possible with the current release.

/ Simon

**Comment by greentaxi-com** [comment]
That's great, I look forward to the release! 

Oddly further up my router config I have multiple request-methods on the same route in the following structure and this doesn't cause the same issue. I initially thought it was the order of the request-methods, but I have played around with the order and haven't got the example above working in a different order.

Router::post('/passenger/', function ($passengerUUID = null) {
    $passenger = new Passenger($passengerUUID);
    $passenger->CreatePassenger();
}, ['defaultParameterRegex' => '[\w\-]+']);

Router::group(['middleware' => GreenTaxi\Core\Authentication::class], function () {

    Router::get('/passenger/{passengerUUID?}/', function ($passengerUUID = null) {
        Authentication::Authorise($passengerUUID);
        $passenger = new Passenger($passengerUUID);
        $passenger->GetPassenger();
    }, ['defaultParameterRegex' => '[\w\-]+']);

    Router::put('/passenger/{passengerUUID?}/', function ($passengerUUID = null) {
        Authentication::Authorise($passengerUUID);
        $passenger = new Passenger($passengerUUID);
        $passenger->UpdatePassenger();
    }, ['defaultParameterRegex' => '[\w\-]+']);

    Router::delete('/passenger/{passengerUUID?}', function ($passengerUUID = null) {
        Authentication::Authorise($passengerUUID);
        $passenger = new Passenger($passengerUUID);
        $passenger->DeletePassenger();
    }, ['defaultParameterRegex' => '[\w\-]+']);

});

**Comment by skipperbent** [comment]
Hi green-taxi.

Well, it can sort of work, but not as intended - it all depends on what order the routes are added and what reqeuest-type they have. However if a route fails (not allowed exception) it will stop trying any of the other routes. I think the route will fail if you place a post route as the first route in your last example and do a `get` request.

In the next release the dash in parameters are supported by default, so you don't need all those regex. More information will be available in the changelog upon release. In the meantime you can remove them and add a single one to your group to simplify your code.

```php
Router::group([
'middleware' => GT\Core\Authentication::class,
'defaultParameterRegex' => '[\w\-]+',
], function () {
   // Routes here no regex needed
}
```

Have a nice day!

/ Simon

**Comment by skipperbent** [comment]
This issue has been resolved in version 4.3.0.0.
Feel free to open another issue if you are still experiencing issues.
